daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/SortedLookupList.scala

```diff
@@ -48,34 +49,34 @@ final class SortedLookupList[+X] private (entries: ImmArray[(String, X)]) extend
 
 object SortedLookupList extends SortedLookupListInstances {
 
-  def fromImmArray[X](entries: ImmArray[(String, X)]): Either[String, SortedLookupList[X]] = {
-    entries.toSeq
-      .groupBy(_._1)
-      .collectFirst {
-        case (k, l) if l.size > 1 => s"key $k duplicated when trying to build map"
-      }
-      .toLeft(new SortedLookupList(entries.toSeq.sortBy(_._1)(Utf8.Ordering).toImmArray))
+  private[this] val EntryOrdering: Ordering[(String, _)] = {
+    case ((key1, _), (key2, _)) => Utf8.Ordering.compare(key1, key2)
   }
 
-  def fromOrderedImmArray[X](
-      entries: ImmArray[(String, X)]): Either[String, SortedLookupList[X]] = {
-    entries
-      .map(_._1)
-      .toSeq
-      .sliding(2)
-      .collectFirst {
-        case Seq(k1, k2) if Utf8.Ordering.gteq(k1, k2) => s"the list $entries is not sorted by key"
-      }
-      .toLeft(new SortedLookupList(entries))
-  }
+  private[this] def nonOrderedEntry[X](entries: ImmArray[(String, X)]): Option[(String, X)] =
+    (entries.iterator zip entries.iterator.drop(1)).collectFirst {
+      case (entry1, entry2) if EntryOrdering.gteq(entry1, entry2) => entry2
+    }
+
+  def fromImmArray[X](entries: ImmArray[(String, X)]): Either[String, SortedLookupList[X]] =
+    nonOrderedEntry(entries.toSeq.sorted(EntryOrdering).toImmArray) match {
+      case None => Right(new SortedLookupList(entries))
+      case Some((key, _)) => Left(s"key $key duplicated when trying to build map")
+    }
+
+  def fromOrderedImmArray[X](entries: ImmArray[(String, X)]): Either[String, SortedLookupList[X]] =
+    nonOrderedEntry(entries) match {
+      case None => Right(new SortedLookupList(entries))
+      case Some(_) => Left(s"the entries $entries is not sorted by key")
+    }
 
   def fromOrderedIterator[X](entries: Iterator[(String, X)]): Either[String, SortedLookupList[X]] =
     fromOrderedImmArray(entries.to[ImmArray])
```

Comment:
I still think the other functions could work on `Iterator[(String, X)]`. If you go back to using `sliding` you only need one.

Reply:
I do not get it. Please feel free to improve my code ;).

Reply:
Hmmm… looks like you can probably just delete this function entirely. It's unused. Does it exist for a reason?

Reply:
Good catch. I was using it in the first version of #7334, but apparently I dropped it.

Reply:
pruned

Reply:
And pushed? 🙃